### ISSUE TITLE
Support of vgo

### ISSUE THREAD

**Issue by xmikus01** [cznic/internal#5]
Package named `internal` can't be currently imported using [vgo](https://github.com/golang/go/wiki/vgo) (issue golang/go#23970). This situation at least prevents importing module `github.com/cznic/lldb` (and transitively `github.com/cznic/ql`) and also intra-dependency import of "github.com/cznic/internal/slice" in [buffer/buffer.go](https://github.com/cznic/internal/blob/4747030f7cf2f4c0a01512b00cd68734b167ac3b/buffer/buffer.go#L36).

Please consider "renaming" this repo (with respect to backward compatibility) or make this package "stable" by integrating it into LLDB (#3).

```
$ mkdir example && cd example
$ echo "module example" > go.mod
$ echo -e "package main\n\nimport _ \"github.com/cznic/ql\"\n\nfunc main () {}" > main.go
$ vgo build
... some successful imports ...
..\..\..\v\github.com\cznic\lldb@v1.1.0\btree.go:16:2: use of internal package not allowed
..\..\..\v\github.com\cznic\lldb@v1.1.0\memfiler.go:13:2: use of internal package not allowed
..\..\..\v\github.com\cznic\internal@v0.0.0-20170905175358-4747030f7cf2\buffer\buffer.go:36:2: use of internal package not allowed
```

**Comment by kenshaw** [comment]
Would also need this for `usql`.

**Comment by cznic** [comment]
> Please consider "renaming" this repo (with respect to backward compatibility) or make this package "stable" by integrating it into LLDB (#3).

Before eventually doing that I'd like to wait for feedback on https://github.com/golang/go/issues/25164#issuecomment-392083411. Also related: cznic/ql#199.

**Comment by kenshaw** [comment]
@cznic a definitive decision on whether or not you're renaming this repository would be appreciated!

May I suggest merging/renaming your various Git repositories? While `vgo` will likely fix this dependency issue, I find that it's best to get ahead of supporting standards. Since it's the intention of Go's core developers to make `vgo` the standard, I might suggest that renaming your repository would be prudent!

**Comment by cznic** [comment]
@kenshaw 

I'm affraid renaming repositories may break builds out there.

I think it would be safer to clone the required repos like `github.com/cznic/internal/buffer` to `github.com/cznic/buffer` and rewrite all affected imports. Those can appear only within other `github.com/cznic/*` packages, so it'll be complete and safe.

Thoughts? 

Once we figure this last detail out I'll do the updates of the affected packages.

**Comment by kenshaw** [comment]
I would suggest renaming `internal` to `common` or something else. As your repositories are the only one using these, I don't believe it'll be an issue -- anytime someone does a `go get -u`, the latest paths would be updated from your dependent repositories. Anything that currently using these repositories will update/build still, as they already have copies on disk. Additionally, if you simply rename `internal` => `common`, I believe `go get -u` for projects with the old names will still work, as I believe GitHub serves the correct path in the `?go-get=1` response. I'll look further into it. Anyway, appreciate your response!

**Comment by kenshaw** [comment]
It's worth noting that according to https://godoc.org/github.com/cznic/internal/buffer?importers there doesn't appear to be any other projects (other than your own) using the `internal/{file,buffer,slice}` packages.

**Comment by kenshaw** [comment]
I just checked manually what happens with a renamed/moved project on GitHub with `go get`:

1) a 301 redirect occurs
2) `go get` sends a `?go-get=1` to the redirected location from 1
3) `git clone` is performed for the URL read from the metadata retrieved in 2

As such, a "lateral" move such that the final package name doesn't change shouldn't affect any existing clones / code on disk.

Anyways, appreciate the consideration! Hope we can get this fixed. Thanks again!

**Comment by cznic** [comment]
OK, please check my line of thoughts.

Consider a fork of, for example, `ql`, owned by , say owner `foo` at `github.com/foo/ql`. Today the owner of that repo can do `go get -u` in any repository that imports the `ql` fork and everything works just fine. If we rename `github.com/cznic/internal` to `github.com/cznic/rename` the forked `ql` build fails so as everything importing that fork.

Except there's the Github redirect feature and even when the `foo`'s fork imports the old, now non-existing `github.com/cznic/internal` package, `go get` will handle the redirect and clones/pulls from the `github.com/cznic/common` repository instead.

The crucial question is: Where does `go get` clone/update the redirected package? If still in the original location, ie. `$GOPATH/src/github.com/cznic/internal` then I agree that it is safe to do the rename. It has even the added advantage that no import paths anywhere have to be updated.

Can you confirm the location on the user disk stays fork-compatible?

And thanks a lot, TIL about Github's redirect feature.

**Comment by xmikus01** [comment]
My suggestions with respect to backward compatibility:
- Do you want to be package `github.com/cznic/internal` reused?
  - If yes, make copy/fork of `github.com/cznic/internal` to `github.com/cznic/common` and rename the package.
  - If no, copy a content of `github.com/cznic/internal` to `github.com/cznic/ql/internal`.
- Refactor `github.com/cznic/ql` to use new package name/path and add `go.mod` file.
- From this new version, vgo is supported for `github.com/cznic/ql` and backward compatibility is maintained.

**Comment by cznic** [comment]
@xmikus01 That was my original plan b/c I didn't know about the Github's redirect feature. Now that @kenshaw brought it to my attention, it seems to be also a non-breaking approach and even simpler. Do you think there's a catch I'm missing? Would love to know about it.

**Comment by kenshaw** [comment]
@cznic I don't believe there would be any problems. Just use the GitHub rename in the settings page.

If you rename this repository to `common` on GitHub, if you later do `go get -u github.com/cznic/internal` the code will be stored on disk at `$GOPATH/src/github.com/cznic/internal`. Then in the  `$GOPATH/src/github.com/cznic/common/.git/config` file, the remote will be defined as: 

```gitconfig
[remote "origin"]
	url = https://github.com/cznic/common
	fetch = +refs/heads/*:refs/remotes/origin/*
```

**Comment by cznic** [comment]
@kenshaw Okay, I'll give it a try. If something goes wrong, it can be reverted. (The hard way: delete the renamed repository completely, create it again and restore its content.)

**Comment by kenshaw** [comment]
You can always just rename it back on GitHub.

**Comment by cznic** [comment]
It does not work, I'll revert in a few moments.

```
jnml@r550:~/tmp> ls -la
celkem 4
drwxr-xr-x  2 jnml users    6  8. čen 17.14 .
drwxr-xr-x 38 jnml users 4096  8. čen 15.59 ..
jnml@r550:~/tmp> export GOPATH=$(pwd)
jnml@r550:~/tmp> go get github.com/cznic/ql
jnml@r550:~/tmp> cd src/github.com/cznic/ql
/home/jnml/tmp/src/github.com/cznic/ql
jnml@r550:~/tmp/src/github.com/cznic/ql> vgo build
vgo: creating new go.mod: module github.com/cznic/ql
vgo: resolving import "github.com/cznic/b"
vgo: finding github.com/cznic/b (latest)
vgo: adding github.com/cznic/b v0.0.0-20180115125044-35e9bbe41f07
vgo: resolving import "github.com/cznic/golex/lex"
vgo: finding github.com/cznic/golex (latest)
vgo: adding github.com/cznic/golex v0.0.0-20170803123110-4ab7c5e190e4
vgo: resolving import "github.com/cznic/lldb"
vgo: finding github.com/cznic/lldb (latest)
vgo: adding github.com/cznic/lldb v1.1.0
vgo: resolving import "github.com/cznic/mathutil"
vgo: finding github.com/cznic/mathutil (latest)
vgo: adding github.com/cznic/mathutil v0.0.0-20180504122225-ca4c9f2c1369
vgo: resolving import "github.com/cznic/strutil"
vgo: finding github.com/cznic/strutil (latest)
vgo: adding github.com/cznic/strutil v0.0.0-20171016134553-529a34b1c186
vgo: finding github.com/cznic/strutil v0.0.0-20171016134553-529a34b1c186
vgo: finding github.com/cznic/mathutil v0.0.0-20180504122225-ca4c9f2c1369
vgo: finding github.com/cznic/lldb v1.1.0
vgo: finding github.com/cznic/golex v0.0.0-20170803123110-4ab7c5e190e4
vgo: finding github.com/cznic/b v0.0.0-20180115125044-35e9bbe41f07
vgo: downloading github.com/cznic/b v0.0.0-20180115125044-35e9bbe41f07
vgo: downloading github.com/cznic/golex v0.0.0-20170803123110-4ab7c5e190e4
vgo: downloading github.com/cznic/lldb v1.1.0
vgo: downloading github.com/cznic/mathutil v0.0.0-20180504122225-ca4c9f2c1369
vgo: downloading github.com/cznic/strutil v0.0.0-20171016134553-529a34b1c186
vgo: resolving import "github.com/cznic/fileutil"
vgo: finding github.com/cznic/fileutil (latest)
vgo: adding github.com/cznic/fileutil v0.0.0-20180108211300-6a051e75936f
vgo: resolving import "github.com/cznic/internal/buffer"
vgo: finding github.com/cznic/internal (latest)
vgo: adding github.com/cznic/internal v0.0.0-20180608151017-b638a8775a65
vgo: resolving import "github.com/cznic/sortutil"
vgo: finding github.com/cznic/sortutil (latest)
vgo: adding github.com/cznic/sortutil v0.0.0-20150617083342-4c7342852e65
vgo: resolving import "github.com/cznic/zappy"
vgo: finding github.com/cznic/zappy (latest)
vgo: adding github.com/cznic/zappy v0.0.0-20160723133515-2533cb5b45cc
vgo: finding github.com/cznic/zappy v0.0.0-20160723133515-2533cb5b45cc
vgo: finding github.com/cznic/sortutil v0.0.0-20150617083342-4c7342852e65
vgo: finding github.com/cznic/internal v0.0.0-20180608151017-b638a8775a65
vgo: finding github.com/cznic/fileutil v0.0.0-20180108211300-6a051e75936f
vgo: downloading github.com/cznic/fileutil v0.0.0-20180108211300-6a051e75936f
vgo: downloading github.com/cznic/internal v0.0.0-20180608151017-b638a8775a65
vgo: downloading github.com/cznic/sortutil v0.0.0-20150617083342-4c7342852e65
vgo: downloading github.com/cznic/zappy v0.0.0-20160723133515-2533cb5b45cc
vgo: resolving import "github.com/edsrzf/mmap-go"
vgo: finding github.com/edsrzf/mmap-go (latest)
vgo: adding github.com/edsrzf/mmap-go v0.0.0-20170320065105-0bce6a688712
vgo: finding github.com/edsrzf/mmap-go v0.0.0-20170320065105-0bce6a688712
vgo: downloading github.com/edsrzf/mmap-go v0.0.0-20170320065105-0bce6a688712
../../../v/github.com/cznic/lldb@v1.1.0/btree.go:16:2: use of internal package not allowed
../../../v/github.com/cznic/lldb@v1.1.0/memfiler.go:13:2: use of internal package not allowed
../../../v/github.com/cznic/internal@v0.0.0-20180608151017-b638a8775a65/buffer/buffer.go:36:2: use of internal package not allowed
jnml@r550:~/tmp/src/github.com/cznic/ql> cd ~/src/golang.org/x/vgo/
jnml@r550:~/src/golang.org/x/vgo> git log -1
jnml@r550:~/src/golang.org/x/vgo> git --no-pager log -1
commit 203abfb0741bf96c7c5e8dab019f6fe9c89bded3 (HEAD -> master, origin/master, origin/HEAD)
Author: Russ Cox <rsc@golang.org>
Date:   Wed Jun 6 16:40:30 2018 -0400

    cmd/go/internal/vgo: fix handling of vendor dir in package patterns
    
    A directory x/vendor containing code is just a package called vendor.
    A directory x/vendor/y is vendored code.
    When a package pattern scans the current module's file tree
    it should include a package called vendor but should not
    include the directories containing vendored code.
    
    Maybe related to a comment on golang/go#25624.
    
    Change-Id: I083a98a9c70c2121cff7a2f394ff985a54bed37a
    Reviewed-on: https://go-review.googlesource.com/116759
    Run-TryBot: Russ Cox <rsc@golang.org>
    TryBot-Result: Gobot Gobot <gobot@golang.org>
    Reviewed-by: Bryan C. Mills <bcmills@google.com>
jnml@r550:~/src/golang.org/x/vgo> 
```

**Comment by kenshaw** [comment]
You'll need to rename the import paths in all the Go projects using this project. Considering your the owner of them all, it shouldn't be a problem I'd hope! I'd be glad to submit PRs for all the projects.

**Comment by cznic** [comment]
Reverted.

@kenshaw But then it means that `foo`'s local fork (as discussed above) works no more. Thanks for the PR offer, but I now believe we should use something like the approach suggested by @xmikus01, which keeps `foo`'s fork working.

To all: I'm quite busy so I didn't thought too much about all of the consequences. I'm not sure I understand all of them and I don't think I'm even aware of all of them. So my current inclination is quite possibly mistaken. Anyway, the safer approach is IMO better even when the other may have some advantage.

I'll have no more time today for experiments. Please comment here anything I should know about before tomorrow, when I hope to create the 'common` repository and change the imports. At least that's what I think/plan now ;-)

**Comment by kenshaw** [comment]
@cznic I do appreciate your attention to this, but any code that is a local fork will continue to work until they rebase as they'll already have the repositories on disk. 

Please consider renaming it again. I'd be glad to offer assistance to anyone who does encounter issues.

Since `vgo` is now the accepted path forward, eventually this project (and all the other projects depending on it) will likely not be `go get`'able in the future. Keep that in mind when weighing your decision! Thanks!

**Comment by xmikus01** [comment]
This issue is now obsolete with closed golang/go#23970. Tested now on go master branch by installing github.com/cznic/ql. The vgo experiment will be replaced by go modules experiment in Go 1.11 which will be released in August 2018. Thereafter this repository can be used as is without any change.
